intermediate_source/reinforcement_q_learning.py

```diff
@@ -433,44 +362,38 @@ def optimize_model():
     # Optimize the model
     optimizer.zero_grad()
     loss.backward()
-    for param in policy_net.parameters():
-        param.grad.data.clamp_(-1, 1)
+    # In-place gradient clipping
+    torch.nn.utils.clip_grad_value_(policy_net.parameters(), 100)
     optimizer.step()
 
 
 ######################################################################
 #
 # Below, you can find the main training loop. At the beginning we reset
-# the environment and initialize the ``state`` Tensor. Then, we sample
-# an action, execute it, observe the next screen and the reward (always
+# the environment and obtain the initial ``state`` Tensor. Then, we sample
+# an action, execute it, observe the next state and the reward (always
 # 1), and optimize our model once. When the episode ends (our model
 # fails), we restart the loop.
 #
-# Below, `num_episodes` is set small. You should download
-# the notebook and run lot more epsiodes, such as 300+ for meaningful
-# duration improvements.
-#
+# Below, `num_episodes` to 1000, but you should the model constantly
+# achieve 500 steps within 600 training episodes.
 
-num_episodes = 50
+num_episodes = 1000
 for i_episode in range(num_episodes):
-    # Initialize the environment and state
-    env.reset()
-    last_screen = get_screen()
-    current_screen = get_screen()
-    state = current_screen - last_screen
+    # Initialize the environment and get it's state
+    state, _ = env.reset()
+    state = torch.tensor(state, dtype=torch.float32, device=device).unsqueeze(0)
     for t in count():
-        # Select and perform an action
         action = select_action(state)
-        _, reward, done, _, _ = env.step(action.item())
+        observation, reward, terminated, truncated, _ = env.step(action.item())
         reward = torch.tensor([reward], device=device)
+        if truncated:
```

Comment:
This is not true, truncation != termination
`done = termination or truncation` in terms of if to reset the environment

This blog post will hopefully be made more public soon https://github.com/pseudo-rnd-thoughts/gym/issues/1

Reply:
Should the replay buffer still be updated on the `truncation` signal with `next_state = torch.tensor(observation, dtype=torch.float32, device=device).unsqueeze(0)`? Performance on this toy problem is similar in either case, but I assume one is more technically correct?

```
done = terminated or truncated
if terminated:
    next_state = None
else:
    next_state = torch.tensor(observation, dtype=torch.float32, device=device).unsqueeze(0)
# Store the transition in memory
memory.push(state, action, next_state, reward)
```

vs

```
done = terminated or truncated
if done:
    next_state = None
else:
    next_state = torch.tensor(observation, dtype=torch.float32, device=device).unsqueeze(0)
# Store the transition in memory
memory.push(state, action, next_state, reward)
```

Reply:
The first one is correct. Thanks for doing that